diff --git a/libsel4vm/src/arch/arm/vgic/gicv2.h b/libsel4vm/src/arch/arm/vgic/gicv2.h
index 1616d63..b25f34d 100644
--- a/libsel4vm/src/arch/arm/vgic/gicv2.h
+++ b/libsel4vm/src/arch/arm/vgic/gicv2.h
@@ -24,6 +24,8 @@
 #define GIC_PADDR   0xc4300000
 #elif defined(CONFIG_PLAT_ZYNQMP)
 #define GIC_PADDR   0xf9000000
+#elif defined(CONFIG_PLAT_BCM2711)
+#define GIC_PADDR   0xff840000
 #else
 #error "Unsupported platform for GIC"
 #endif
diff --git a/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/device_map.h b/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/device_map.h
new file mode 100644
index 0000000..d1ebc81
--- /dev/null
+++ b/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/device_map.h
@@ -0,0 +1,7 @@
+/*
+ * Copyright 2019, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
diff --git a/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/devices.h b/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/devices.h
new file mode 100644
index 0000000..d1ebc81
--- /dev/null
+++ b/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/devices.h
@@ -0,0 +1,7 @@
+/*
+ * Copyright 2019, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
diff --git a/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/guest_vcpu_util.h b/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/guest_vcpu_util.h
new file mode 100644
index 0000000..82a8bf3
--- /dev/null
+++ b/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/guest_vcpu_util.h
@@ -0,0 +1,7 @@
+/*
+ * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#define PLAT_CPU_COMPAT "arm,cortex-a72"
diff --git a/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/vpci.h b/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/vpci.h
new file mode 100644
index 0000000..80c7345
--- /dev/null
+++ b/libsel4vmmplatsupport/plat_include/bcm2711/sel4vmmplatsupport/plat/vpci.h
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2019, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+#pragma once
+
+/* PCI host bridge memory regions are defined in the pci dts node
+ * supplied to the Linux guest. These values are also reflected here.
+ */
+
+/* PCI host bridge configration space */
+#define PCI_CFG_REGION_ADDR 0x41000000
+/* PCI host bridge IO space */
+#define PCI_IO_REGION_ADDR 0x40000000
+/* Size of PCI configuration space */
+#define PCI_CFG_REGION_SIZE 0x1000000
+/* Size of PCI IO space  */
+#define PCI_IO_REGION_SIZE 0x10000
+/* PCI memory space */
+#define PCI_MEM_REGION_ADDR 0x42000000ull
+/* PCI memory space size */
+#define PCI_MEM_REGION_SIZE 0x1000000
+
+/* FDT IRQ controller address cells definition */
+#define GIC_ADDRESS_CELLS 0x2
